### PR TITLE
Add distinct minimizer heatmap for Kraken2

### PR DESCRIPTION
**Pull request by maxibor**
Since version 2.1.0, Kraken2 integrated the distinct minimizer count ability of [KrakenUniq](https://genomebiology.biomedcentral.com/articles/10.1186/s13059-018-1568-0)
This new function comes with a slightly altered report (see [Kraken2 docs](https://github.com/DerrickWood/kraken2/wiki/Manual#distinct-minimizer-count-information)) format.

This PR adds a heatmap of the duplication rate of the minimizers (i.e. kmers) which is very useful to detect read stacking, and false positive hits.

- When encountering a "old" report, it behaves as before
- When encountering a "new report" (with distinct minimizer count) it adds the duplication heatmap.

PR checklist:

- [x] This comment contains a description of changes (with reason)
- [x] `CHANGELOG.md` has been updated

### CODE REVIEW

**Comment by maxibor** [comment]
Screenshot 2021-02-23 at 18 29 58

**Comment by maxibor** [comment]
I didn't see it before, but this should solve #1379

**Comment by ewels** [comment]
I could have sworn I added Black CI to MultiQC, but I can't see it now. Could you please run `black` on your code?

Also, please some test data to https://github.com/ewels/MultiQC_TestData/tree/master/data/modules/kraken for this.

Otherwise, LGTM from quick read of the code 👍🏻 Will review again once we have test data.

**Comment by aeu79** [comment]
@maxibor Testing with my data, nice plots! (now I need to figure them out)

BTW, the unclassified disappeared from the top 5 taxa plot. Was it intended?
image
image
And that is, of course, changing the percentages.

**Comment by aeu79** [comment]
Another issue is that it crashes when the reports are mixed (I mean, std and with minimizers in the same folder). Sorry, I was being picky with my tests for #1379 :smile:

**Comment by maxibor** [comment]
@aeu79 adc7087 should fix these issues
- The unclassified reads were brought back thanks to a `f.seek(0)` (reading the file to check the 'log version' was consuming the first line)
- The mixed reports are now nicely handled


Screenshot 2021-02-24 at 14 35 13


@ewels This is now black formatted

**Comment by maxibor** [comment]
I've added test data with this PR https://github.com/ewels/MultiQC_TestData/pull/187

**Comment by aeu79** [comment]
Guauu, that was fast, congratulations! Looks great:
image
I reached to the same conclusion by comparing the reads and unique minimizers:
image
(the one in red from your plot doesn't even show up in the top 50 taxa when I order them by minimizers)

BTW, with many files I'm getting this warning (only with your version):
```
version:  1.10.dev0 (adc7087)
All std
[INFO   ]         multiqc : This is MultiQC v1.10.dev0 (adc7087)
[INFO   ]         multiqc : Template    : default
[INFO   ]         multiqc : Searching   : /home/agu/Desktop/testMQC
[INFO   ]          kraken : Found 104 reports
/home/agu/GitHub/multiqc/MultiQC/multiqc/plots/bargraph.py:486: UserWarning: FixedFormatter should only be used together with FixedLocator
  axes.set_xticklabels(["{:.0f}%".format(x) for x in vals])
[WARNING]          kraken : Kraken2 reports of different versions were found
```

**Comment by ewels** [comment]
Merge conflicts after merging #1347 - will take a look now.

**Comment by ewels** [comment]
ok, conflicts fixed and code working. I can see that the TestData sample `metagenome.kreport` now has 100% `Unclassified` reads in the General Stats table (instead of empty cells) and that it's reads are now _Other_ instead of _Unclassified_ for some of the levels in the Top Taxa plot. But the new heatmap looks kind of uninspiring for me 😓 

Screenshot 2021-03-04 at 22 39 22

* Is this just that the test data is not super useful or am I doing something wrong?
* If it's possible to generate an empty plot like this, can we check for that and skip it if not useful?
* If it is the test data, please could we add some more interesting test data into the TestData repo?

Thanks all!

Phil

**Comment by ewels** [comment]
> BTW, with many files I'm getting this warning (only with your version):
```
version:  1.10.dev0 (adc7087)
All std
[INFO   ]         multiqc : This is MultiQC v1.10.dev0 (adc7087)
[INFO   ]         multiqc : Template    : default
[INFO   ]         multiqc : Searching   : /home/agu/Desktop/testMQC
[INFO   ]          kraken : Found 104 reports
/home/agu/GitHub/multiqc/MultiQC/multiqc/plots/bargraph.py:486: UserWarning: FixedFormatter should only be used together with FixedLocator
  axes.set_xticklabels(["{:.0f}%".format(x) for x in vals])
[WARNING]          kraken : Kraken2 reports of different versions were found
```

@aeu79 - this warning was unrelated to the changes in this PR. It's a warning from matplotlib due to the core MultiQC code. It's only just started appearing due to an update in their end. I suspect it only showed for you here because you were running with more samples, triggering MultiQC to generate flat image bargraphs instead of interactive.

Anyway, it should now be fixed in `master` in de0f01b5db483c64ebba45e26e3984743a5e26c5 so you should hopefully not see it again. Thanks for reporting 👍🏻

File: multiqc/modules/kraken/kraken.py [hunk 1]
```diff
@@ -295,9 +384,70 @@ def top_five_barplot(self):
...
+                    counts_shown[s_name] = 0
+                for row in d:
+                    if row["rank_code"] == rank_code:
+                        if row["classif"] == classif:
+                            if classif not in rank_data[s_name]:
+                                rank_data[s_name][classif] = 0
+                            try:
+                                rank_data[s_name][classif] = row["minimizer_duplication"]
+                            except KeyError:
+                                del rank_data[s_name]
```
Comment (maxibor) [inline multiqc/modules/kraken/kraken.py:433]: Updated duplication plot to remove samples from kraken report without distinct minimizer information

**Comment by maxibor** [comment]
- Updated test data here: https://github.com/ewels/MultiQC_TestData/pull/189

**Comment by ewels** [comment]
Thanks @maxibor! I brought back my check to avoid repeated noticed about mixed reports, which had gotten lost.

I'm running tests on the data in MultiQC_TestData > `data/modules/kraken` and after your update the report now crashes with a JavaScript error (fairly difficult to do usually!).

I'll take a look now, but generally speaking it's good to also try running on just this data as then you'll come across these same issues as me and can probably save some time 😉

**Comment by ewels** [comment]
Ok added an extra loop in c2faaeb51ea91fcc0f73b8546a1e8ae173c7ad3b to strip out any heatmap samples with zero values. This fixes the report for me and I now get a heatmap with two cells. I also tweaked the heatmap config as I figure it doesn't need to be square (the x- and y-categories are not the same) and I also specified that the xcats are not sample names, so that they are not affected by hiding / renaming / highlighting in the report.

kraken-topfive-duplication_plot

I'm happy with this so will merge as soon as the tests are passing - thank you! 👍🏻  (especially for speedy response after the long delay).

Please check that you're ok with my changes and let me know ASAP if not as we can fix in a follow-up PR.

Phil

**Comment by ewels** [comment]
@aeu79 - note that your updated regex from #1383 is now in this PR also, so should hopefully address your comment: https://github.com/ewels/MultiQC/pull/1383#issuecomment-790630204

Shout if there's anything missing still.

Phil

**Pull request merged**